io.openems.edge.battery.fenecon.home/src/io/openems/edge/battery/fenecon/home/FeneconHomeBatteryImpl.java

```diff
@@ -294,111 +330,286 @@ protected ModbusProtocol defineModbusProtocol() throws OpenemsException {
+								m(generateBcuChannel(i, "_SOH"), new UnsignedWordElement(towerOffset + 9), // [%]
+										ElementToChannelConverter.SCALE_FACTOR_MINUS_1), //
+								m(generateBcuChannel(i, "_VOLTAGE"), new UnsignedWordElement(towerOffset + 10), // [V]
+										ElementToChannelConverter.SCALE_FACTOR_MINUS_1), //
+								m(generateBcuChannel(i, "_CURRENT"), new UnsignedWordElement(towerOffset + 11), // [A]
+										ElementToChannelConverter.SCALE_FACTOR_MINUS_1), //
+								m(generateBcuChannel(i, "_MIN_CELL_VOLTAGE"),
+										new UnsignedWordElement(towerOffset + 12)), // [mV]
+								m(generateBcuChannel(i, "_MAX_CELL_VOLTAGE"),
+										new UnsignedWordElement(towerOffset + 13)), // [mV]
+								m(generateBcuChannel(i, "_AVARAGE_CELL_VOLTAGE"),
+										new UnsignedWordElement(towerOffset + 14)), //
+								m(generateBcuChannel(i, "_MAX_CHARGE_CURRENT"),
+										new UnsignedWordElement(towerOffset + 15)), //
+								m(generateBcuChannel(i, "_MIN_CHARGE_CURRENT"),
+										new UnsignedWordElement(towerOffset + 16)), //
+								m(generateBcuChannel(i, "_BMS_SERIAL_NUMBER"),
+										new UnsignedWordElement(towerOffset + 17)), //
+								m(generateBcuChannel(i, "_NO_OF_CYCLES"), new UnsignedWordElement(towerOffset + 18)), //
+								m(new UnsignedWordElement(towerOffset + 19)) //
+										.m(generateBcuChannel(i, "_DESIGN_CAPACITY"),
+												ElementToChannelConverter.SCALE_FACTOR_MINUS_1) // [Ah]
+										.m(Battery.ChannelId.CAPACITY, ElementToChannelConverter.DIRECT_1_TO_1) // [%]
+										.build(), //
+								m(generateBcuChannel(i, "_USABLE_CAPACITY"), new UnsignedWordElement(towerOffset + 20), //
+										ElementToChannelConverter.SCALE_FACTOR_MINUS_1), // [Ah]
+								m(generateBcuChannel(i, "_REMAINING_CAPACITY"),
+										new UnsignedWordElement(towerOffset + 21), //
+										ElementToChannelConverter.SCALE_FACTOR_MINUS_1), // [Ah]
+								m(generateBcuChannel(i, "_MAX_CELL_VOLTAGE_LIMIT"),
+										new UnsignedWordElement(towerOffset + 22)), //
+								m(generateBcuChannel(i, "_MIN_CELL_VOLTAGE_LIMIT"),
+										new UnsignedWordElement(towerOffset + 23))));
+			}
+		} catch (OpenemsException e) {
+			this.log.info("Dynamic Channels could not created");
+		}
+
+		this.channelMap = this.createCellVoltAndTempDynamicChannels(numberOfTowers, numberOfModulePerTower);
+		try {
+			// Register is available -> add Registers for current hardware to protocol
+			int offset = ModuleParameters.ADDRESS_OFFSET_FOR_CELL_VOLT_AND_TEMP.getValue();
+			int voltOffset = ModuleParameters.VOLTAGE_ADDRESS_OFFSET.getValue();
+			int voltSensors = ModuleParameters.VOLTAGE_SENSORS_PER_MODULE.getValue();
+			for (int t = 1; t <= numberOfTowers; t++) {
+				String towerString = "TOWER_" + t + "_OFFSET";
+				int towerOffset = ModuleParameters.valueOf(towerString).getValue();
+				for (int i = 1; i < numberOfModulePerTower + 1; i++) {
+					Collection<AbstractModbusElement<?>> elements = new ArrayList<>();
```

Comment:
You could also directly initialize the Array here, as you already know the length. It's not wrong of course, though.